rclcpp/src/rclcpp/utilities.cpp

```diff
@@ -277,15 +311,121 @@ typedef void (* signal_handler_t)(int);
+#include <windows.h>
+#elif defined(__APPLE__)
+#include <dispatch/dispatch.h>
+#else  // posix
+#include <semaphore.h>
+#endif
+
+static std::atomic_bool g_wait_for_signal_is_setup = ATOMIC_VAR_INIT(false);
+#if defined(_WIN32)
+HANDLE g_signal_handler_sem;
+#elif defined(__APPLE__)
+dispatch_semaphore_t g_signal_handler_sem;
+#else  // posix
+sem_t g_signal_handler_sem;
+#endif
 
-/// Condition variable for timed sleep (see sleep_for).
-static std::condition_variable g_interrupt_condition_variable;
-/// Mutex for protecting the global condition variable.
-static std::mutex g_interrupt_mutex;
+void
+rclcpp::setup_wait_for_signal()
+{
+#if defined(_WIN32)
+  g_signal_handler_sem = CreateSemaphore(
+    NULL,  // default security attributes
+    0,  // initial semaphore count
+    1,  // maximum semaphore count
+    NULL);  // unnamed semaphore
+  if (NULL == g_signal_handler_sem) {
+    throw std::runtime_error("CreateSemaphore() failed in setup_wait_for_signal()");
+  }
+#elif defined(__APPLE__)
+  g_signal_handler_sem = dispatch_semaphore_create(0);
+#else  // posix
+  if (-1 == sem_init(&g_signal_handler_sem, 0, 0)) {
+    throw std::runtime_error(std::string("sem_init() failed: ") + strerror(errno));
+  }
+#endif
+  g_wait_for_signal_is_setup.store(true);
+}
+
+void
+rclcpp::teardown_wait_for_signal() noexcept
+{
+  if (!g_wait_for_signal_is_setup.exchange(false)) {
+    return;
+  }
+#if defined(_WIN32)
+  CloseHandle(g_signal_handler_sem);
+#elif defined(__APPLE__)
+  dispatch_release(g_signal_handler_sem);
+#else  // posix
+  if (EINVAL == sem_destroy(&g_signal_handler_sem)) {
+    RCLCPP_ERROR(g_logger, "invalid semaphore in teardown_wait_for_signal()");
+  }
+#endif
+}
+
+void
+rclcpp::wait_for_signal() noexcept
+{
+  if (!g_wait_for_signal_is_setup.load()) {
+    RCLCPP_ERROR(g_logger, "called wait_for_signal() before setup_wait_for_signal()");
+    return;
+  }
+#if defined(_WIN32)
+  DWORD dw_wait_result = WaitForSingleObject(g_signal_handler_sem, INFINITE);
+  switch (dw_wait_result) {
+    case WAIT_ABANDONED:
+      RCLCPP_ERROR(
+        g_logger, "WaitForSingleObject() failed in wait_for_signal() with WAIT_ABANDONED: %s",
+        GetLastError());
+      break;
+    case WAIT_OBJECT_0:
+      // successful
+      break;
+    case WAIT_TIMEOUT:
+      RCLCPP_ERROR(g_logger, "WaitForSingleObject() timedout out in wait_for_signal()");
+      break;
+    case WAIT_FAILED:
+      RCLCPP_ERROR(
+        g_logger, "WaitForSingleObject() failed in wait_for_signal(): %s", GetLastError());
+      break;
+    default:
+      RCLCPP_ERROR(
+        g_logger, "WaitForSingleObject() gave unknown return in wait_for_signal(): %s",
+        GetLastError());
+  }
+#elif defined(__APPLE__)
+  dispatch_semaphore_wait(g_signal_handler_sem, DISPATCH_TIME_FOREVER);
+#else  // posix
+  int s;
+  do {
+    s = sem_wait(&g_signal_handler_sem);
+  } while (-1 == s && EINTR == errno);
+#endif
+}
+
+void
+rclcpp::notify_signal_handler() noexcept
```

Comment:
@wjwwood same consideration about notifying the caller of an error condition here.

Reply:
In this case, this function cannot fail and there's really no point in returning something here, since it would just mean I do `if (failed) // same thing I'm doing in the function now, i.e. RCLCPP_ERROR`.